# -*- org-confirm-babel-evaluate: nil -*-
#+TITLE: ParkinsonPathAI (Working title) - AI Models for Pathway Analysis in Parkinson’s Disease
#+OPTIONS: ^:nil
#+Author: Sam Neaves
#+HTML_HEAD: <style>
#+HTML_HEAD: pre.example { max-height: 300px; overflow-y: auto; border: 1px solid #ccc; padding: 10px; background-color: #f9f9f9; }
#+HTML_HEAD: </style>


* Introduction

This project focuses on analyzing single-cell transcriptomic data from
brain cells of both control and Parkinson’s disease donors. Our goal
is to characterize the complex interplay between the RAS and dopamine
pathways, we will attempt to do this using symbolic AI techniques. By employing methods like
descriptive rule induction, causal analysis, and model
building/repair, we aim to identify and understand unique cell states
associated with Parkinson’s disease.

Our approach reconciles gene networks with experimental data, allowing
us to reason about network consistency and potentially discover key
molecular signatures involved in disease progression. This repository will
contain code, data preprocessing steps, and model implementation
details for this exploratory analysis.

This is an open scientific project, actively developed in the
open-source community and a continuous work in progress. As the
results are preliminary, they may evolve over time with further
analysis and validation. If you spot any errors or have suggestions,
please feel free to open an issue. Contributions and feedback from the
community are welcome to help improve and refine our understanding of
the RAS and dopamine pathways in Parkinson’s disease.

* Code Organization

This project is hosted on GitHub and follows a clear and structured
approach to organizing code and documentation.

** Code Accessibility and Literate Programming
The code is written in a
[[https://en.wikipedia.org/wiki/Literate_programming][literate programming style]] , integrating code
and explanatory prose. This approach enhances readability and provides
clear context for each step in the analysis. The primary medium for
this documentation is [[https://orgmode.org/][Emacs Org mode files]] , which resemble Markdown
but offer several additional features, including:

- Advanced Formatting: Support for tables, source code blocks, and
  inline evaluation of code.
  
- Integration with Analysis: Directly execute embedded code blocks.
  
- Export Capabilities: Generate various formats like HTML, PDF, and
  plain text from the same source file.
  
- Tangled Scripts and Workflow Execution
From these Org files, tangled scripts are generated. Tangling extracts
the source code blocks into standalone scripts, enabling flexible
execution:


- Step-by-Step Execution: Run individual scripts to focus on specific
  parts of the analysis. This is especially useful if you are
  interested in a single stage of the workflow.
  
- Pipeline Execution: The tangled scripts are also integrated into a
  Nextflow pipeline, allowing the entire workflow to be executed in
  one process. The Nextflow execution enhances reproducibilty,
  portabilty and scalabilty of the code.
  

** File Formats and Visualization
The Org files are plain text and can be viewed or edited in any text
editor. However, to improve accessibility and provide a visually
appealing presentation, the Org files are exported to HTML and hosted
on GitHub Pages. This HTML version enhances the user experience by
incorporating multimedia elements like images, links, and interactive
visualizations.


** Accessing the Files
- Org File: The main Org file for this project is located at: [Link to Org file].
- HTML View: The corresponding HTML version can be accessed here: [Link to GitHub Pages].
This organization ensures that the code is both transparent and
user-friendly, making it easy for collaborators to understand,
replicate, and extend the analysis.


* Data

Biological cell samples were collected postmortem from the ventral
substantia nigra of seven control samples, nine Parkinson's disease
samples, and four samples from patients with Incidental Lewy Body
Disorder. The nuclei of these cells were extracted and sequenced,
totaling 23,885 nuclei. The sequecning was deep and full
length. [INSERT DETALS OF SEQUENCING TECH OR REFERNCE]

** System biology approach

Our approach to this analysis is grounded in symbolic systems
biology. In systems biology, our goal is to understand and model
biological reality at a systems level. This means focusing on how
multiple aspects of biology interact to produce observable outcomes,
contrasting with traditional approaches that isolate individual
phenomena. Symbolic systems biology differs from quantitative systems
biology in the level of abstraction it applies.


Using discrete symbols to represent components of the system, rather
than specific quantities, offers several advantages. First,
abstracting measured data into symbolic form allows us to manage noise
in the data. Second, this abstraction reduces the size of the state
space we analyze, making certain computational methods more
feasible. Third, our goal is not only to understand the system but
also to control it. Therefore, it’s essential that our models and
their dynamics are interpretable, allowing suggested interventions and
their potential consequences to be validated in a laboratory setting.


Finally, there is evidence that symbolic abstraction is sufficient to
model biological dynamics in many cases.


Biological reality is highly complex. While the central dogma—DNA
transcribed to RNA, translated to protein—outlines the basic process
of cellular maintenance and control, the specifics of transcription,
translation, and metabolic pathways leading to disease are intricate
and require further exploration.


For instance, in somatic cells, DNA remains stable as cells
replicate. Differences between cells arise through processes such as
differentiation and gene regulation. Transcription can occur in
bursts, and alternative splicing allows for multiple proteins from a
single gene. Additionally, protein complexes may form with the aid of
chaperones, and epigenetic mechanisms, such as methylation and histone
modification, affect gene accessibility.




*** Intreptation of sequence data as instances in state space.

In our analysis of transcriptome from nuclei, previous work has
identified cell types using x method.
In addition cell states have been identified by y method.

** RAS MAPK pathway

In this work we are intrested in these pathways and the interplay.

I obtained two lists of genes from the two pathways from a collegue
(Viola - check where she got them from).
In order to get a full list of genes that we are intreseted in and their
interactions, we augment our intital list and query pathway commons to
obtain a network of genes. We use the pyBravo tool to make this query.

The intial list is :

Raas genes:
#+begin_src bash :results output :exports both
  cat ./data/initial_gene_list_files/raas_genes.txt
#+end_src

#+RESULTS:
#+begin_example
ACE
ADCY
AGTR1
AGTR2
AGT
ATF2
CCL2
CCL5
DAG1
ELK1
ERK1
ERK2
FOS
GNAQ
GRB2
IP3
JAK2
JNK
JUN
MAP2K1
MAP2K2
MAP2K4
MAP3K1
NFKB1
NO
NOX1
MAPK
PAK
PGE2
PI3K
PIP2
PKA
PKC
PLCG1
PTGER2
PTK2
PTK2B
PTPN6
RAC1
RAF1
HRAS
KRAS
REN
SHC1
SOS1
STAT1
STAT3
TNF
#+end_example

Dopamine KEGG pathway:

#+begin_src bash :results output :exports both
  cat ./data/initial_gene_list_files/DopamineKEGGpathway.txt
#+end_src

#+RESULTS:
#+begin_example
AKT3
CREB3
GNB5
ADCY5
COMT
CREB1
ATF2
ATF6B
MAPK14
CREB3L4
CALML6
DDC
DRD1
DRD2
DRD3
DRD4
DRD5
AKT1
AKT2
LRTOMT
PLCB1
FOS
GNAI1
GNAI2
GNAI3
GNAL
GNAQ
GNAO1
GNAS
GNB1
GNB2
GNB3
GNG3
GNG4
GNG5
GNG7
GNG10
GNG11
GNGT1
GNGT2
PPP2R3B
GRIA1
GRIA2
GRIA3
GRIA4
GRIN2A
GRIN2B
GSK3A
GSK3B
ITPR1
ITPR2
ITPR3
KCNJ3
KCNJ5
KCNJ6
KCNJ9
KIF5A
KIF5B
KIF5C
ARNTL
ARRB1
ARRB2
MAOA
MAOB
ATF4
CALY
GNG13
CALML5
PLCB2
PLCB3
PLCB4
GNG2
PPP1CA
PPP1CB
PPP1CC
PPP2R3C
PPP2CA
PPP2CB
PPP2R1A
PPP2R1B
PPP2R2A
PPP2R2B
PPP2R2C
PPP2R3A
PPP2R5A
PPP2R5B
PPP2R5C
PPP2R5D
PPP2R5E
PPP3CA
PPP3CB
PPP3CC
PRKACA
PRKACB
PRKACG
PRKCA
PRKCB
PRKCG
PPP2R2D
GNG12
MAPK8
MAPK11
MAPK13
GNB4
MAPK12
SCN1A
CREB3L2
SLC6A3
TH
CACNA1A
CACNA1B
CACNA1C
CACNA1D
CALM1
CALM2
CALM3
CALML3
CAMK2A
CAMK2B
CAMK2D
CAMK2G
PPP1R1B
CREB3L3
CREB3L1
CALML4
GNG8
CLOCK
CREB5
SLC18A2
SLC18A1
MAPK9
MAPK10

#+end_example

These were provided by Viola.

The intersection of these files is:
#+begin_src bash :results output :exports both
  grep -Fxf ./data/initial_gene_list_files/DopamineKEGGpathway.txt ./data/initial_gene_list_files/raas_genes.txt
#+end_src

#+RESULTS:
: ATF2
: FOS
: GNAQ

The union of these files is:

#+begin_src bash :results output :export both
  cat ./data/initial_gene_list_files/DopamineKEGGpathway.txt ./data/initial_gene_list_files/raas_genes.txt | sort | uniq | tee ./data/initial_gene_list_files/initial_gene_list.txt
  wc -l ./data/initial_gene_list_files/initial_gene_list.txt
#+end_src

#+RESULTS:
#+begin_example

ACE
ADCY
ADCY5
AGT
AGTR1
AGTR2
AKT1
AKT2
AKT3
ARNTL
ARRB1
ARRB2
ATF2
ATF4
ATF6B
CACNA1A
CACNA1B
CACNA1C
CACNA1D
CALM1
CALM2
CALM3
CALML3
CALML4
CALML5
CALML6
CALY
CAMK2A
CAMK2B
CAMK2D
CAMK2G
CCL2
CCL5
CLOCK
COMT
CREB1
CREB3
CREB3L1
CREB3L2
CREB3L3
CREB3L4
CREB5
DAG1
DDC
DRD1
DRD2
DRD3
DRD4
DRD5
ELK1
ERK1
ERK2
FOS
GNAI1
GNAI2
GNAI3
GNAL
GNAO1
GNAQ
GNAS
GNB1
GNB2
GNB3
GNB4
GNB5
GNG10
GNG11
GNG12
GNG13
GNG2
GNG3
GNG4
GNG5
GNG7
GNG8
GNGT1
GNGT2
GRB2
GRIA1
GRIA2
GRIA3
GRIA4
GRIN2A
GRIN2B
GSK3A
GSK3B
HRAS
IP3
ITPR1
ITPR2
ITPR3
JAK2
JNK
JUN
KCNJ3
KCNJ5
KCNJ6
KCNJ9
KIF5A
KIF5B
KIF5C
KRAS
LRTOMT
MAOA
MAOB
MAP2K1
MAP2K2
MAP2K4
MAP3K1
MAPK
MAPK10
MAPK11
MAPK12
MAPK13
MAPK14
MAPK8
MAPK9
NFKB1
NO
NOX1
PAK
PGE2
PI3K
PIP2
PKA
PKC
PLCB1
PLCB2
PLCB3
PLCB4
PLCG1
PPP1CA
PPP1CB
PPP1CC
PPP1R1B
PPP2CA
PPP2CB
PPP2R1A
PPP2R1B
PPP2R2A
PPP2R2B
PPP2R2C
PPP2R2D
PPP2R3A
PPP2R3B
PPP2R3C
PPP2R5A
PPP2R5B
PPP2R5C
PPP2R5D
PPP2R5E
PPP3CA
PPP3CB
PPP3CC
PRKACA
PRKACB
PRKACG
PRKCA
PRKCB
PRKCG
PTGER2
PTK2
PTK2B
PTPN6
RAC1
RAF1
REN
SCN1A
SHC1
SLC18A1
SLC18A2
SLC6A3
SOS1
STAT1
STAT3
TH
TNF
     178 ./data/initial_gene_list_files/initial_gene_list.txt
#+end_example

Next in order to understand which other genes have known interactions
with this list and how this list is related we query pathway commons
using pybravo.

For Pybravo to function, it must connect to a SPARQL endpoint from
Pathway Commons. Unfortunately, this resource can be unreliable at
times. However, it’s possible to set up a local SPARQL endpoint by
building a Docker container. A Docker recipe file is available in this
GitHub repository:

https://github.com/mkarikom/PathwayCommons-VOS#pathwaycommons-v12-on-virtuoso-7

With Pybravo installed and an appropriate Conda environment set up, we
can configure Pybravo to connect to our local instance of the Pathway
Commons SPARQL endpoint for querying.


We have set up the command to run from a Biomake file (makeprog in the
scripts directory).



#+begin_src bash :dir "./scripts/" :exports both :eval query :results output
  biomake "augment_gene_list" | tail -n 40
#+end_src

#+RESULTS:
#+begin_example
--- Upstream regulation network in 68.57 seconds ---
Number of nodes = 1339
Number of edges = 12032
SIF network written to ../data/pybravo_output/expanded_reg.sif
Basic regulation reaction provenance written to ../data/pybravo_output/expanded_reg-provenance.csv

| Node | Degree Centrality |
|------|------|
| NOG | 0.264 | 
| SP1 | 0.182 | 
| LEF1 | 0.17 | 
| MAZ | 0.168 | 
| TCF3 | 0.155 | 
| FOXO4 | 0.143 | 
| TGFB1 | 0.138 | 
| JUN | 0.136 | 
| MYC | 0.128 | 
| PAX4 | 0.123 | 

--- Network simplification in 2.49 seconds ---
SIF network written to ../data/pybravo_output/expanded_reg-unified.sif
Basic regulation reaction provenance written to ../data/pybravo_output/expanded_reg-unified-provenance.csv
Nodes after simplification = 1229
Edges after simplification = 11912

| Node | Degree Centrality |
|------|------|
| NOG | 0.276 | 
| DAND5 | 0.195 | 
| LEF1 | 0.182 | 
| MAZ | 0.178 | 
| TCF7L1 | 0.176 | 
| FOXO4 | 0.155 | 
| JUN | 0.151 | 
| TGFB1 | 0.149 | 
| HNF1A | 0.149 | 
| PVT1 | 0.139 | 


Target augment_gene_list built
#+end_example




This results in the creation of four files in the pybravo_output dir

expanded_reg_md10-provenance.csv
expanded_reg_md10-unified-provenance.csv expanded_reg_md10-unified.sif
expanded_reg_md10.sif

The expanded_reg_md10-unified.sif file can be viewed in cyto scape.

#+CAPTION: Gene Regulation Network
#+NAME:   fig:1
[[./data/plots/expanded_reg_md10-unified.png]]






** Data info

"Just sending the single nuclei data for the dopaminergic neurons in our atlas as Caleb suggested.
The file “SNatlas_DaNs_seurat.RData” is in the usual Dropbox folder.
The Seurat object in it is called “sn_atlas_dans” and contains all
neurons across all disease stages and subtypes. You can access disease
by sn_atlas_dans$Disease and cell subtype by
sn_atlas_dans$CellSubType. And extract gene expression data as raw
counts genes_counts<-as.matrix(GetAssayData(sn_atlas_dans, slot =
"counts")) or normalised
genes_norm<-as.matrix(GetAssayData(sn_atlas_dans, slot = "data"))."



#+begin_src R :session :eval query 
  library(dplyr)
  library(Seurat)
  library(patchwork)

  data_location <- "/Users/samneaves/Documents/test_data/forSam/SNatlas_DaNs_seurat.RData"
  load(data_location)

#+end_src

#+RESULTS:
: sn_atlas_dans

#+begin_src R :session
head(sn_atlas_dans$CellSubType)
#+end_src

#+RESULTS:
| DaN_0 |
| DaN_0 |
| DaN_1 |
| DaN_1 |
| DaN_0 |
| DaN_0 |

*** Get the normalised expression values
#+begin_src R :session :results none
genes_norm<-as.matrix(GetAssayData(sn_atlas_dans, slot = "data"))
#+end_src

#+RESULTS:


*** Get the count expression values

#+begin_src R :session :results none
genes_counts<-as.matrix(GetAssayData(sn_atlas_dans, slot = "counts"))
#+end_src

** Finding the overlap of list of genes that we get back from pathway commons and our data
#+begin_src R :session :results output :exports both
  network_data <- read.table("./data/pybravo_output/expanded_reg_md10-unified.sif", quote = "", header = FALSE, stringsAsFactors = FALSE, sep = "\t")
  head(network_data)
  nodes <- unique(c(network_data$V1, network_data$V3))
#+end_src

#+RESULTS:
:         V1         V2     V3
: 1 Fra1/JUN ACTIVATION   CCL2
: 2 Fra1/JUN ACTIVATION   Fra1
: 3 Fra1/JUN ACTIVATION    IL6
: 4 Fra1/JUN ACTIVATION p14ARF
: 5 Fra1/JUN ACTIVATION   MMP9
: 6 Fra1/JUN ACTIVATION  CXCL8

#+begin_src R :session :results output :exports both
  length(nodes)
#+end_src

#+RESULTS:
: [1] 1053

#+begin_src R :session :exports both
  # Convert the rownames of genes_counts into a vector
  genes_names <- rownames(genes_counts)

  # Get the intersection between the 'name' column in node_table and the rownames of genes_counts_subset_df
  common_names <- intersect(nodes, genes_names)

  # View the result
  length(common_names)
#+end_src

#+RESULTS:
: 526


In the graph representation we have genes and protein complexes.
A protein complex node is represented by Item/Item. For example

SMAD3/SMAD4/JUN/FOS

When we match our gene expression data to the network. These nodes
will be unobserved. We will need to infer if they are able to
form/activate.

In the way that the pybravo paper describes these nodes are used in
there iggy analysis. The 'part_of' edge is treated as plus. But I think the standard consistency
check is only looking for at least one positive edge to label a node
as consistant So not sure if that is appropiate. (will add detail to
this point later).

For the time being the number of complexes seem to be :

#+begin_src R :session :exports both
# Identify rows in the 'name' column that contain '/'
complexes <- grep("/", nodes)

# Count the number of complexes
num_complexes <- length(complexes)

# View the count
num_complexes
#+end_src

#+RESULTS:
: 296

#+begin_src R :session :exports both
# Count occurrences of each edge type in the second column (V2)
edge_counts <- table(network_data$V2)

# Display the counts of each edge type
print(edge_counts)
#+end_src

#+RESULTS:
| ACTIVATION | 1710 |
| INHIBITION |  655 |
| PART_OF    |  567 |
| UNKNOWN    |   95 |


#+begin_src R :session :exports both
  str(sn_atlas_dans)
  summary(sn_atlas_dans)
#+end_src

#+RESULTS:

We want to select Cells that are either control or parkinsons
We want to only select the genes that are in our network.
#+begin_src R :session :exports both
  sn_atlas_dans_ctrl_vs_pd <- subset(sn_atlas_dans, subset = Disease %in% c("CTR", "PD_B5-6", "PD_B3-4"))
  sn_atlas_dans_ctrl_vs_pd_pathway <- sn_atlas_dans_ctrl_vs_pd[common_names,]
  dim(sn_atlas_dans_ctrl_vs_pd_pathway)

  
#+end_src

#+RESULTS:
|  526 |
| 6385 |

#+begin_src R :session :exports both

# Extract the expression matrix for the subset
expression_matrix <- as.data.frame(t(as.matrix(sn_atlas_dans_ctrl_vs_pd_pathway@assays$RNA@data)))

# Add Disease column from metadata
expression_matrix$Disease <- sn_atlas_dans_ctrl_vs_pd_pathway@meta.data$Disease

# Save to CSV
write.csv(expression_matrix, file = "./data/cell_expression_with_disease.csv", row.names = TRUE)


#+end_src

#+RESULTS:


#+begin_src R :session :exports both

# Extract the count matrix for the subset and transpose it for cells as rows
count_matrix <- as.data.frame(t(as.matrix(sn_atlas_dans_ctrl_vs_pd_pathway@assays$RNA@counts)))

# Add Disease column from metadata
count_matrix$Disease <- sn_atlas_dans_ctrl_vs_pd_pathway@meta.data$Disease

# Save to CSV
write.csv(count_matrix, file = "./data/cell_count_with_disease.csv", row.names = TRUE)

#+end_src

#+RESULTS:



* Lets make an up set plot

#+begin_src R :session :exports both

  
#+end_src

#+begin_src R :file 4.png :session :results graphics file :exports both
library(lattice)
xyplot(1:10 ~ 1:10)
#+end_src

#+RESULTS:
[[file:4.png]]

#+begin_src R :file upset_demo.png :session :results graphics file :exports both
  movies <- read.csv( system.file("extdata", "movies.csv", package = "UpSetR"), header=T, sep=";" )
  library("UpSetR")
  upset(movies,attribute.plots=list(gridrows=60,plots=list(list(plot=scatter_plot, x="ReleaseDate", y="AvgRating"),
  list(plot=scatter_plot, x="ReleaseDate", y="Watches"),list(plot=scatter_plot, x="Watches", y="AvgRating"),
  list(plot=histogram, x="ReleaseDate")), ncols = 2))
#+end_src

#+RESULTS:
[[file:upset_demo.png]]

Try and have the ids in the first colum
#+begin_src python :session myorangesession :results output :exports both
import pandas as pd

# Extract the features and class labels from data
X = data_with_class_discrete.X  # feature matrix
Y = data_with_class_discrete.Y  # class labels

# Get instance IDs
instance_ids = [str(data_with_class_discrete.metas[i][0]) for i in range(len(Y))]

# Initialize a DataFrame to store coverage information
coverage_df = pd.DataFrame(index=range(X.shape[0]), columns=['id'] + [str(rule) for rule in classifier.rule_list])

# Fill the 'id' column with instance IDs
coverage_df['id'] = instance_ids

# Evaluate each rule and fill coverage DataFrame
for rule in classifier.rule_list:
    covered_examples = rule.evaluate_data(X)  # returns a boolean array of covered examples
    
    # Fill the DataFrame with 1s and 0s based on rule coverage
    coverage_df[str(rule)] = covered_examples.astype(int)

# Save the coverage DataFrame to a CSV file
coverage_df.to_csv('./data/rule_coverage.csv', index=False)

print("Coverage matrix saved to './data/rule_coverage.csv'")
#+end_src

#+RESULTS:
: Coverage matrix saved to 'rule_coverage.csv'

#+begin_src R :file upset_instances.png :session :results graphics file :exports both
  rule_cover_data <- read.csv("./data/rule_coverage.csv", row.names = 1)
   library(UpSetR)

  data_binary <- data.frame(t(rule_cover_data))

  upset(data_binary, nsets = 20, sets = colnames(data), keep.order = TRUE)
#+end_src

#+RESULTS:
[[file:upset_instances.png]]


#+begin_src R :file upset_rules.png :session :results graphics file :exports both
  #rule_cover_data <- read.csv("./data/rule_coverage.csv", row.names = 1)
  #library(UpSetR)

  data_binary <- data.frame(rule_cover_data)

  upset(data_binary, nsets = 20, sets = colnames(data), keep.order = TRUE)
#+end_src

#+RESULTS:
[[file:upset_rules.png]]

#+begin_src R :file upset_rules_complex.png :session :results graphics file :exports both
  #rule_cover_data <- read.csv("./data/rule_coverage.csv", row.names = 1)
  #library(UpSetR)
  #detach("package:UpSetR", unload=TRUE)
  library(ComplexUpset)

  set_size(8, 3)
  genres = colnames(movies)[18:24]
  upset(movies, genres, name='genre', width_ratio=0.1)
#+end_src

#+RESULTS:
[[file:upset_rules_complex.png]]

* Mini graph for each rule?
* thing
#+header: :width 2000 :height 800
#+begin_src R :file complex_upset_rules.png :session :results graphics file :exports both



  library(ggplot2)
  library(ComplexUpset)
  movies = as.data.frame(ggplot2movies::movies)
  head(movies, 3)
  genres = colnames(movies)[18:24]
  genres

  movies[genres] = movies[genres] == 1
  t(head(movies[genres], 3))

  movies[movies$mpaa == '', 'mpaa'] = NA
  movies = na.omit(movies)

  #upset(movies, genres, name='genre', width_ratio=0.1)

  rule_cover_data <- read.csv("./data/rule_coverage.csv", row.names = 1)

  items = colnames(rule_cover_data)[1:10]


#  quartz(width = 100, height = 5)
#pdf("upset_plot_test_10_8.pdf", width = 100, height = 8)
upset(rule_cover_data, items, name='genre', width_ratio=0.1, n_intersections=15)
#dev.off()
#+end_src

#+RESULTS:
[[file:complex_upset_rules.png]]

