# -*- org-confirm-babel-evaluate: nil -*-
#+TITLE: ParkinsonPathAI (Working title) - AI Models for Pathway Analysis in Parkinson’s Disease

#+Author: Sam Neaves

* Introduction

This project focuses on analyzing single-cell transcriptomic data from
brain cells of both control and Parkinson’s disease donors. Our goal
is to characterize the complex interplay between the RAS and dopamine
pathways, we will attempt to do this using symbolic AI techniques. By employing methods like
descriptive rule induction, causal analysis, and model
building/repair, we aim to identify and understand unique cell states
associated with Parkinson’s disease.

Our approach reconciles gene networks with experimental data, allowing
us to reason about network consistency and potentially discover key
molecular signatures involved in disease progression. This repository will
contain code, data preprocessing steps, and model implementation
details for this exploratory analysis.

This is an open scientific project, actively developed in the
open-source community and a continuous work in progress. As the
results are preliminary, they may evolve over time with further
analysis and validation. If you spot any errors or have suggestions,
please feel free to open an issue. Contributions and feedback from the
community are welcome to help improve and refine our understanding of
the RAS and dopamine pathways in Parkinson’s disease.

* Data

Biological cell samples were collected postmortem from the ventral
substantia nigra of seven control samples, nine Parkinson's disease
samples, and four samples from patients with Incidental Lewy Body
Disorder. The nuclei of these cells were extracted and sequenced,
totaling 23,885 nuclei. The sequecning was deep and full
length. [INSERT DETALS OF SEQUENCING TECH OR REFERNCE]

** System biology approach

Our approach to this analysis is based on symbolic systems biology.
In systems biology we aim to understand and model the biological
reality at a system level, that is understanding how many aspects of
biology interact to cause our observations. This is in conrast to
looking at specific decompatialised observations. Symbolic system
biology differs from quantitive symbolic biology in the level of
abstraction.
The use of discrete symbols to represent aspects of the system rather
than specific quantitites has a number of advantages which we seek to
exploit.
Firstly, the abstraction of measured data to symbols is a way to
handle noise.
Secondly, the abstraction reduces the size of state space that we are examining.
This means that certain computational methods become tractable.
Thirdly, as our aim is to understand what is happening and ultimatly
control the system.  It is therefore important that our models and their dynamics
are intertplble, so that suggested interventions and there
consequences can be ratified in a laboratory setting.

Forthyly, there is evidence that symbolic abstraction is sufficent to model the dynamics in
many cases.

Biological reality is extremely complicated.

The basic idea of DNA -> (transcribed) RNA -> (translated) Protein,
which takes part in a series of biochemical reactions to build,
maintain and controll cells is
sound and present however the details around transcription,
translation and metabolic pathways ultimatly leading to disease are
complex and what we seek to understand.

In somatic cells the DNA is immutable as they replicate into daughter
cells.
The differnce between cells .
Cell differnation.
Gene regulatory networks.

Transcription can be bursty.
Alternative splicing
Protein complex formation. - chaperones.
Epigentics, methylation and histone access.


